File: app/src/main/java/com/orgzly/android/util/QueryUtils.java
package com.orgzly.android.util;

import com.orgzly.android.query.Condition;

public class QueryUtils {
    public static String extractFirstBookNameFromQuery(Condition condition) {
        if (condition instanceof Condition.And) {
            Condition.And c = (Condition.And) condition;

            for (Condition innerCondition : c.getOperands()) {
                String result = extractFirstBookNameFromQuery(innerCondition);

                if (result != null) {
                    return result;
                }
            }
        }

        if (condition instanceof Condition.InBook) {
            Condition.InBook c = (Condition.InBook) condition;

            if (!c.getNot()) {
                return c.getName();
            }
        }

        return null;
    }
}

File: app/src/main/java/com/orgzly/android/query/QueryUtils.kt
package com.orgzly.android.query

object QueryUtils {
    @JvmStatic
    fun extractFirstBookNameFromQuery(condition: Condition?): String? {
        if (condition is Condition.And) {
            val (operands) = condition

            for (innerCondition in operands) {
                val result = extractFirstBookNameFromQuery(innerCondition)

                if (result != null) {
                    return result
                }
            }

        } else if (condition is Condition.Or) {
            val (operands) = condition

            for (innerCondition in operands) {
                val result = extractFirstBookNameFromQuery(innerCondition)

                if (result != null) {
                    return result
                }
            }

        } else if (condition is Condition.InBook) {
            val (name, not) = condition

            if (!not) {
                return name
            }
        }

        return null
    }
}
